lib/assets/javascripts/utils/validation.js

```diff
@@ -0,0 +1,223 @@
+    case 'url':
+      return validator.isValidUrl(value);
+    case 'password':
+      return validator.isValidPassword(value);
+    case 'radio':
+      return validator.isValidText(value);
+    case 'select':
+    case 'checkbox':
+      return validator.isValidText(value);
+    case 'js-combobox':
+      return validator.isValidText(value);
+    default:
+      return false;
+  }
+};
+
+const getDefaultValidationMessage = (type) => {
+  switch (type) {
+    case 'text':
+      return constants.VALIDATION_MESSAGE_TEXT;
+    case 'textarea':
+      return constants.VALIDATION_MESSAGE_TEXT;
+    case 'number':
+      return constants.VALIDATION_MESSAGE_NUMBER;
+    case 'email':
+      return constants.VALIDATION_MESSAGE_EMAIL;
+    case 'url':
+      return constants.VALIDATION_MESSAGE_URL;
+    case 'password':
+      return constants.VALIDATION_MESSAGE_PASSWORD;
+    case 'radio':
+      return constants.VALIDATION_MESSAGE_RADIO;
+    case 'checkbox':
+      return constants.VALIDATION_MESSAGE_CHECKBOX;
+    case 'js-combobox':
+      return constants.VALIDATION_MESSAGE_SELECT;
+    default:
+      return constants.VALIDATION_MESSAGE_DEFAULT;
+  }
+};
+const getValidationMessage = (el) => {
+  if ($(el).attr('data-validation-error')) {
+    return $(el).attr('data-validation-error');
+  }
+  // Use the default validation error message if none was specified
+  return getDefaultValidationMessage(getValidationTypeForElement(el));
+};
+const valid = (el) => {
+  $(el).parent().removeClass(validationStates.hasError);
+  $(el).attr(ariaInvalid(false));
+  $(el).next().hide();
+};
+const invalid = (el) => {
+  $(el).parent().addClass(validationStates.hasError);
+  $(el).attr(ariaInvalid(true));
+  $(el).next().show();
+};
+const addValidationMessage = (el) => {
+  if (isString($(el).attr('id'))) {
+    const id = $(el).attr('id');
+    if (!isString($(el).attr('aria-describedby'))) {
+      $(el).after(blockHelp(`help-${id}`, getValidationMessage(el)));
+      $(el).attr('aria-describedby', `help-${id}`);
+      $(el).attr('data-validatable', 'true');
+    }
+  }
+};
+const removeValidationMessage = (el) => {
+  if (isString($(el).attr('id'))) {
+    $(el).parent().find('.help-block').remove();
+    $(el).removeAttr('aria-describedby');
+    $(el).removeAttr('data-validatable');
+  }
+};
+const isRequired = (ctx) => {
+  if (isObject(ctx)) {
+    return ($(ctx).attr('aria-required') && $(ctx).attr('aria-required') === 'true');
+  }
+  return false;
+};
+const checkValidations = (el) => {
+  const type = getValidationTypeForElement(el);
+  const value = getValue(type, el);
+
+  // If the field is required or it has a value
+  if (isRequired(el) || isString(value)) {
+    if (isValid(type, value)) {
+      valid(el);
+    } else {
+      invalid(el);
+      return false;
+    }
+  }
+  return true;
+};
+
+export const isValidatable = (ctx) => {
+  if (isObject(ctx)) {
+    return isString($(ctx).attr('data-validatable'));
```

Comment:
why we adding data-validatable?

Reply:
ah function is leftover from earlier. I guess linter didn't throw an 'unused const' error because its being exported.